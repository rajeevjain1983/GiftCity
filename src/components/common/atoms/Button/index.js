// 9fbef606107a605d69c0edbcd8029e5d 
export { default } from './views';
export { BUTTON_VARIATION, BUTTON_FILL } from './Button.constants';
