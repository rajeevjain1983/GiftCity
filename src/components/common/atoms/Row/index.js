// 9fbef606107a605d69c0edbcd8029e5d 
// @flow
export { default } from './views';
