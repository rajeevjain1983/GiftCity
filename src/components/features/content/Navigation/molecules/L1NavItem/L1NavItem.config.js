// 9fbef606107a605d69c0edbcd8029e5d 
export const DELAY_TO_OPEN = 100;

export const CLEARANCE_ID = '505518';

export default { DELAY_TO_OPEN, CLEARANCE_ID };
