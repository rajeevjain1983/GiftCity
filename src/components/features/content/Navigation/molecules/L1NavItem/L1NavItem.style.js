
import { css } from 'styled-components';
export default css`
  /* @media ${props => props.theme.mediaQuery.large} { */
    .nav-bar-item-label {
      width: 100%;
      display: inline-block;
    }
    &.is-open {
      .nav-bar-l2 {
        display: block;
      }
      .nav-bar-l1-content{
        .icon-arrow{
          background: url(/icons/carrot-small-down.png) no-repeat;
        }
      }
    }
  .icon-arrow{
    background: url(/icons/carrot-small-rights.png) no-repeat;
    width: 10px;
    height: 10px;
    display: inline-block;
    }

    

    .nav-bar-l1-content{
      display:flex;
    }

  /* } */
`;
