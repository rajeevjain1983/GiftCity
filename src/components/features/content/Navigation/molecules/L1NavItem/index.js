export { default } from './L1NavItem';
