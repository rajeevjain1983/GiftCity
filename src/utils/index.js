
import {
  disableBodyScroll as disableBodyScrollLib,
  enableBodyScroll as enableBodyScrollLib,
  // clearAllBodyScrollLocks,
} from 'body-scroll-lock';
import { breakpoints} from '../../styles/themes/mediaQuery';

export const getViewportInfo = () => {
    if (typeof window === 'undefined') return null;
    if (window.viewportInfo) return window.viewportInfo;
  
    const width = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
    const height = Math.max(document.documentElement.clientHeight, window.innerHeight || 0);
    const isMobile = width < parseInt(breakpoints.medium, 10);
    const isTablet = !isMobile && width < parseInt(breakpoints.large, 10);
    const isDesktop = !isMobile && !isTablet;
  
    window.viewportInfo = {
      width,
      height,
      isMobile,
      isTablet,
      isDesktop,
    };
  
    return window.viewportInfo;
  };


 

  export const isMobileApp = () => {
    return typeof navigator !== 'undefined' && navigator.product === 'ReactNative';
  };

  /**
 * To Identify whether the device is ios for web.
 */

export const isiOSWeb = () => {
    const userAgent =
      typeof navigator !== 'undefined' && (navigator.userAgent || navigator.vendor || window.opera);
    if (/iPad|iPhone|iPod/.test(userAgent) && !window.MSStream) {
      return true;
    }
    return false;
  };
  
  export function isClient() {
    return typeof window !== 'undefined' && !isMobileApp();
  }
  /**
   * To Identify whether the device is Android for web.
   */
  
  export const isAndroidWeb = () => {
    const userAgent =
      typeof navigator !== 'undefined' && (navigator.userAgent || navigator.vendor || window.opera);
    if (/Android/.test(userAgent)) {
      return true;
    }
    return false;
  };
  
  /**
   * To Identify whether the device is Android for web.
   */
  export const isMobileWeb = () => {
    return (isClient() && isiOSWeb()) || (isClient() && isAndroidWeb());
  };
  
  /**
 * Enable Body Scroll, Moving it to common utils and putting a check of Mobile app at one place instead of containers.
 */
export const enableBodyScroll = targetElem => {
  if (isClient()) {
    if (isiOSWeb() && targetElem) {
      enableBodyScrollLib(targetElem);
      return;
    }
    const [body] = document.getElementsByTagName('body');
    body.classList.remove('disableBodyScroll');
  }
};

/**
 * Disable Body Scroll
 */
export const disableBodyScroll = targetElem => {
  if (isClient()) {
    if (isiOSWeb() && targetElem) {
      disableBodyScrollLib(targetElem);
      return;
    }
    const [body] = document.getElementsByTagName('body');
    body.classList.add('disableBodyScroll');
  }
};

export const getIconPath = iconName => {
  return `/icons/${iconName}`;
}