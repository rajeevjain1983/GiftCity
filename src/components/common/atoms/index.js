// 9fbef606107a605d69c0edbcd8029e5d 

import Row from './Row';
import Col from './Col';
// import Anchor from './Anchor';
// import Button from './Button';
// import DamImage from './DamImage';
// import DynamicTag from './DynamicTag';
// import Image from './Image';
// import RichText from './RichText';
// import TextBox from './TextBox';
// import Heading from './Heading';
import BodyCopy from './BodyCopy';
// import SelectBox from './Select';
// import LabeledRadioButton from './LabeledRadioButton';
// import TextItems from './TextItems';
// import InputCheckBox from './InputCheckbox';
// import StoreStaticMap from './StoreStaticMap';
// import Skeleton from './Skeleton';
// import BackToTop from './BackToTop';
// import PLPSkeleton from './PLPSkeleton';
// import DLPSkeleton from './DLPSkeleton';
// import FavoriteSkeleton from './FavoriteSkeleton';
// import Spinner from './Spinner';
// import TextArea from './TextArea';
// import AppBanner from './AppBanner';
// import DynamicModule from './DynamicModule';

export {
  Col,
  Row,
//   Anchor,
//   Button,
  
//   DamImage,
//   DynamicTag,
//   Image,
//   InputCheckBox,
//   RichText,
//   TextBox,
//   Heading,
  BodyCopy,
//   SelectBox,
//   LabeledRadioButton,
//   TextItems,
//   StoreStaticMap,
//   Skeleton,
//   BackToTop,
//   PLPSkeleton,
//   DLPSkeleton,
//   Spinner,
//   TextArea,
//   FavoriteSkeleton,
//   AppBanner,
//   DynamicModule,
};
