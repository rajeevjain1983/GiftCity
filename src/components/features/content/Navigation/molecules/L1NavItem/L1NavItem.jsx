// 9fbef606107a605d69c0edbcd8029e5d
import React from "react";
import Link from "next/link";
import Router from "next/router";
import withStyles from "../../../../../common/hoc/withStyle";
import { BodyCopy } from "../../../../../common/atoms";
import { getViewportInfo } from "../../../../../../utils";
import { KEY_CODES } from "../../../../../../constants/keyboard.constants";
import style from "./L1NavItem.style";
import { DELAY_TO_OPEN } from "./L1NavItem.config";

const HideDrawerContext = React.createContext({});
const HideDrawerProvider = HideDrawerContext.Provider;
export const HideDrawerConsumer = HideDrawerContext.Consumer;

/**
 * This function highlights clearance links in red color on the base of id in unbxd
 * @param {*} id
 */
const highlightContent = (id) => {
  return id === "505518" || id === "454010" ? `highlighted` : ``;
};

class L1NavItem extends React.PureComponent {
  constructor() {
    super();
    this.state = {
      hovered: false,
    };
    this.timeOutHandler = null;
  }

  componentWillUnmount() {
    document.body.removeEventListener("keydown", this.handleDocumentKeydown);
  }

  /**
   * This function handles document keydown event when L2 drawer is opened
   */
  handleDocumentKeydown = (e) => {
    const { dataLocator } = this.props;
    if (
      e.keyCode === KEY_CODES.KEY_TAB ||
      (e.keyCode === KEY_CODES.KEY_TAB && e.shiftKey) ||
      e.keyCode === KEY_CODES.KEY_ENTER ||
      e.keyCode === KEY_CODES.KEY_SPACE
    ) {
      clearTimeout(this.timeOutHandler);
      // close drawer when focused element is not part of current opened drawer
      if (
        !document.getElementById(dataLocator).contains(document.activeElement)
      ) {
        this.onMouseLeave();
      }
    }
  };

  /**
   * This function will be used to open the l2 link items
   */
  onHover = (e) => {
    e.preventDefault();
    e.stopPropagation();
    if (getViewportInfo().isDesktop) {
      clearTimeout(this.timeOutHandler);
      this.timeOutHandler = setTimeout(() => {
        this.setHoveredState(true);
      }, DELAY_TO_OPEN);
    }
  };

  /**
   * This function toggle hovered state of drawer and bind/unbind keydown
   * @param {Boolean} isHovered
   * @param {*} onClick
   */
  setHoveredState = (isHovered, onClick = () => {}) => {
    this.setState(
      {
        hovered: isHovered,
      },
      () => {
        onClick();
      }
    );
    if (isHovered && getViewportInfo().isDesktop) {
      document.body.addEventListener("keydown", this.handleDocumentKeydown);
    } else {
      document.body.removeEventListener("keydown", this.handleDocumentKeydown);
    }
  };

  /**
   * This function will be used to close the l2 links wrapper when user click on any link inside it.
   */

  hideL2Nav = () => {
    this.setHoveredState(false);
  };

  /**
   * This function will be used to close the l2 link items
   */
  onMouseLeave = () => {
    if (getViewportInfo().isDesktop) {
      clearTimeout(this.timeOutHandler);
      this.setHoveredState(false);
    }
  };

  /**
   * This function will be used to open the l2 link items on Enter/Space Press
   */
  onKeyPressed = (e) => {
    if (
      (getViewportInfo().isDesktop && e.keyCode === KEY_CODES.KEY_ENTER) ||
      e.keyCode === KEY_CODES.KEY_SPACE
    ) {
      e.preventDefault();
      this.setHoveredState(true);
    }
  };

  handleFocusBlur = () => {
    const timer = setTimeout(() => {
      document.activeElement.blur();
      clearTimeout(timer);
    }, 0);
    Router.events.off("routeChangeComplete", this.handleFocusBlur);
  };

  /**
   * This function handles if navigation drawer needs to open on current viewport or now
   * @param {*} onClick
   */
  openNavigationDrawer = (hasL2) => (e) => {
    const { onClick } = this.props;
    if (!getViewportInfo().isDesktop && hasL2) {
      e.preventDefault();
      e.stopPropagation();
      this.setHoveredState(true, onClick);
    } else if (getViewportInfo().isDesktop) {
      e.stopPropagation();
      clearTimeout(this.timeOutHandler);
      this.setHoveredState(false);
      Router.events.on("routeChangeComplete", this.handleFocusBlur);
    }
  };

  /**
   * To render the l1 content\
   */
  renderL1Content = (classForRedContent, displayName) => {
    return (
      <div className="nav-bar-l1-content">
        <span className={`nav-bar-item-label ${classForRedContent}`}>
          {displayName}
        </span>

        <span className="icon-arrow" />
      </div>
    );
  };

  render() {
    const {
      name,
      displayName,
      className,
      dataLocator,
      index,
      children,
      removeL1Focus,
      // hasL2,
      // clickData,
      // sizesRange,
      // linkOverrideUrl,
      ...others
    } = this.props;

    const { hovered } = this.state;

    let classForHovered = "";
    if (hovered) {
      classForHovered = "is-open";
      this.childRendered = true;
    }
    // If we receive flag showOnlyOnApp then we add this class to links to hide them

    // This class is used to highlight link in red color, it performs check based on id
    const classForRedContent = highlightContent(name);

    return (
      <React.Fragment>
        <HideDrawerProvider
          value={{
            hideL2Nav: this.hideL2Nav,
          }}
        >
          <BodyCopy
            component="li"
            className={`${className} ${classForHovered} nav-bar-l1-item  `}
            // fontFamily="secondary"
            fontSize={["fs13", "fs13", "fs15"]}
            fontWeight="semibold"
            // color="text.primary"
            // lineHeight="lh115"
            data-locator={dataLocator}
            onMouseEnter={this.onHover}
            onMouseLeave={this.onMouseLeave}
            id={dataLocator}
            role="menuitem"
            {...others}
          >
            <Link
              href={`c/${name}`}
              onKeyDown={this.onKeyPressed}
              onClick={this.openNavigationDrawer(true)}
              aria-expanded={!!classForHovered}
            >
              {this.renderL1Content(classForRedContent, displayName, index)}
            </Link>
            {(hovered || this.childRendered) && children}
            {/* {hovered ? (
              <div
                className={`${className} l1-overlay ${classForHovered}`}
                onMouseEnter={this.onMouseLeave}
              />
            ) : null} */}
          </BodyCopy>
        </HideDrawerProvider>
      </React.Fragment>
    );
  }
}

export { L1NavItem as L1NavItemVanilla };
export default withStyles(L1NavItem, style);
