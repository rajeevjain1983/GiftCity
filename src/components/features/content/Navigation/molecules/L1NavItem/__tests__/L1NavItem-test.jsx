// 9fbef606107a605d69c0edbcd8029e5d 
import React from 'react';
import { shallow } from 'enzyme';
import { getViewportInfo } from '@tcp/core/src/utils';
import { BodyCopy } from '@tcp/core/src/components/common/atoms';
import { DELAY_TO_OPEN } from '../L1NavItem.config';
import { L1NavItemVanilla as L1NavItem } from '../L1NavItem';

jest.mock('@tcp/core/src/utils', () => {
  const originalModule = jest.requireActual('@tcp/core/src/utils');
  return {
    ...originalModule,
    getViewportInfo: jest.fn(),
  };
});

const data1 = {
  categoryContent: {
    name: '',
    description: '',
    seoToken: '',
    mainCategory: {
      promoBadge: [
        {
          text: '30% off shoes',
          style: 'style1',
        },
      ],
    },
    dataLocator: 'dataLocator',
  },
};

const data2 = {
  categoryContent: {
    name: 'Girl',
    seoToken: '',
    description: 'Sizes 4-14',
    mainCategory: {
      promoBadge: '',
    },
  },
};

describe('Drawer component', () => {
  it('renders correctly', () => {
    const L1NavItemComp = shallow(<L1NavItem {...data1} />);

    expect(L1NavItemComp).toMatchSnapshot();
  });

  it('has description loaded', () => {
    const L1NavItemComp = shallow(<L1NavItem {...data2} />);

    expect(L1NavItemComp.find('.nav-bar-item-label')).toHaveLength(1);
  });

  it('set hovered state to true on mouseenter', () => {
    getViewportInfo.mockImplementation(() => ({ isDesktop: true }));

    const L1NavItemComp = shallow(<L1NavItem {...data2} />);
    L1NavItemComp.find(BodyCopy).simulate('mouseEnter', {
      preventDefault: () => {},
      stopPropagation: () => {},
    });
    setTimeout(() => {
      expect(L1NavItemComp.state('hovered')).toBe(true);
    }, DELAY_TO_OPEN);
  });

  it('set hovered state to true on mouseleave', () => {
    getViewportInfo.mockImplementation(() => ({ isDesktop: true }));

    const L1NavItemComp = shallow(<L1NavItem {...data2} />);
    L1NavItemComp.find(BodyCopy).simulate('mouseLeave');

    expect(L1NavItemComp.state('hovered')).toBe(false);
  });
});
