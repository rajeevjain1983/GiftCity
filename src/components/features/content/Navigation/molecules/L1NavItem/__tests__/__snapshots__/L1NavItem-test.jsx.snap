// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer component renders correctly 1`] = `
<Fragment>
  <ContextProvider
    value={
      Object {
        "hideL2Nav": [Function],
      }
    }
  >
    <Styled(BodyCopy)
      className="undefined  nav-bar-l1-item"
      color="text.primary"
      component="li"
      customStyle={Object {}}
      data-locator=""
      fontFamily="secondary"
      fontSize={
        Array [
          "fs13",
          "fs13",
          "fs15",
        ]
      }
      fontWeight="semibold"
      id=""
      letterSpacing="normal"
      lineHeight="lh115"
      link={null}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      textAlign="left"
    >
      <Styled(Anchor)
        aria-expanded={false}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          className="nav-bar-l1-content"
        >
          <span
            className="nav-bar-item-label "
          />
          <span
            className="nav-bar-item-content "
            data-locator="promo_badge_undefined"
          >
            <Styled(PromoBadge)
              data={
                Array [
                  Object {
                    "style": "style1",
                    "text": "30% off shoes",
                  },
                ]
              }
            />
          </span>
          <span
            className="icon-arrow"
          />
        </div>
      </Styled(Anchor)>
    </Styled(BodyCopy)>
  </ContextProvider>
</Fragment>
`;
